pkg/function/server/server.go

```diff
@@ -26,35 +26,54 @@ func New() *server {
 
 // RegisterMapper registers the map operation handler to the server.
 // Example:
-// func handle(ctx context.Context, key string, data functionsdk.Datum) functionsdk.Messages {
-// 	_ = data.EventTime() // Event time is available
-// 	_ = data.Watermark() // Watermark is available
-// 	return functionsdk.MessagesBuilder().Append(functionsdk.MessageToAll(data.Value()))
-// }
 //
-// func main() {
-// 	server.New().RegisterMapper(functionsdk.MapFunc(handle)).Start(context.Background())
-// }
+//	func handle(ctx context.Context, key string, data functionsdk.Datum) functionsdk.Messages {
+//		_ = data.EventTime() // Event time is available
+//		_ = data.Watermark() // Watermark is available
+//		return functionsdk.MessagesBuilder().Append(functionsdk.MessageToAll(data.Value()))
+//	}
+//
+//	func main() {
+//		server.New().RegisterMapper(functionsdk.MapFunc(handle)).Start(context.Background())
+//	}
 func (s *server) RegisterMapper(m functionsdk.MapHandler) *server {
 	s.svc.Mapper = m
 	return s
 }
 
+// RegisterMapperT registers the mapT operation handler to the server.
+// Example:
+//
+//	func handle(ctx context.Context, key string, data functionsdk.Datum) functionsdk.MessageTs {
+//		_ = data.EventTime() // Event time is available
```

Comment:
I am thinking how we will get `EventTime` and `Watermark` here if we are to use this new handler to get event-time and then assign watermark.

Reply:
My understanding is that within the input datum, although we already have EventTime and Watermark, whether or not they are valid, is up to the user to determine.

User can implement the UDF to completely ignore the passed in EventTime and Watermark and assign an irrelevant new EventTime. User can also choose to consider the passed in EventTime as valid and maybe generate a new EventTime based on the passed-in one(plus one minute if it makes sense).

At the end of the UDF execution, the only thing numaflow cares about is the `functionsdk.MessageTs` returned.

Reply:
fair point, please document this.